pyblazing/pyblazing/apiv2/context.py

```diff
@@ -3236,3 +3214,169 @@ def log(self, query, logs_table_name="bsql_logs"):
             self.logs_initialized = True
 
         return self.sql(query)
+
+    def _get_progress_bar_format(self):
+        pbfmt = "Steps Complete {n_fmt}/{total_fmt}|{bar}|{percentage:3.0f}% ({elapsed} elapsed)"
+        return pbfmt
+
+    def _wait_completed_single_node(self, graph):
+        query_complete = False
+        while not query_complete:
+            query_complete = graph.query_is_complete()
+            sleep(0.005)
+
+    def _wait_completed_distributed(self, ctxToken):
+        query_complete = False
+        while not query_complete:
+            dask_futures = []
+            for node in self.nodes:
+                worker = node["worker"]
+                dask_futures.append(
+                    self.dask_client.submit(
+                        getQueryIsComplete, ctxToken, workers=[worker], pure=False
+                    )
+                )
+            workers_is_complete = self.dask_client.gather(dask_futures)
+            query_complete = all(workers_is_complete)  # all workers returned true
+            sleep(0.005)
+
+    def _run_progress_bar_single_node(self, graph):
+        from tqdm import tqdm
+
+        query_complete = False
+
+        progress = graph.get_progress()
+        thepdf = queryProgressAsPandas(progress)
+        themax = len(thepdf)
+        batches_completed = thepdf["batches_completed"].sum()
+        thepdf = thepdf.drop(thepdf[~thepdf.finished].index)
+        thesteps = len(thepdf)
+
+        pbar = tqdm(
+            total=themax,
+            miniters=1,
+            bar_format=self._get_progress_bar_format(),
+            leave=False,
+        )
+        pbar2 = tqdm(
+            miniters=1, bar_format="Total Batches Processed: {n_fmt}", leave=False
+        )
+
+        pbar.update(thesteps)
+        pbar2.update(batches_completed)
+        last = thesteps
+        last_sum_batches = batches_completed
+        while True:
+            query_complete = graph.query_is_complete()
+            progress = graph.get_progress()
+            pdf = queryProgressAsPandas(progress)
+            batches_completed = pdf["batches_completed"].sum()
```

Comment:
I think this could be refactored so that we dont have the same logic in multiple places. Something like this:

def update_pbars(pdf, pbar, pbar2, last_steps, last_sum_batches):
             batches_completed = pdf["batches_completed"].sum()
            pdf = pdf.drop(pdf[~pdf.finished].index)
            thesteps = len(pdf)
            if last != thesteps:
                delta = thesteps - last
                last = thesteps
                pbar.update(delta)
                if last_sum_batches != batches_completed:
                    delt2 = batches_completed - last_sum_batches
                    pbar2.update(delt2)
                    last_sum_batches = batches_completed
             return [last_steps, last_sum_batches]